## Linear Regression Building Homework 

We want you to build a linear regression model to predict the selling_price of a car in both in Python and in R.  

Do some basic data exploration in Python (you may wish to use pandas_profiling for this if it works for you)  


In [1]:
import pandas as pd 

cars = pd.read_csv("car_dekho_details.csv") 

In [2]:
cars 
cars = cars.drop('name', 1) 

In [3]:
import pandas_profiling 

In [4]:
#pandas_profiling.ProfileReport(cars) 

In [5]:
cars['year'] = cars['year'].astype('category') 

In [6]:
cars = pd.get_dummies(cars, drop_first = True) 

In [7]:
from sklearn.linear_model import LinearRegression 

In [8]:
y = cars["selling_price"]  
x = cars.drop(columns = "selling_price") 

In [9]:
model = LinearRegression() 

In [10]:
model.fit(x, y) 

LinearRegression()

In [11]:
model.score(x, y) 

0.4852283948297037

In [12]:
model.intercept_ 

1095829.2274971162

In [13]:
model.coef_ 

array([-4.25282343e-01,  8.05726169e+03, -2.05140129e+05, -3.82411681e+05,
       -2.90245299e+04, -4.67939962e+04, -4.80785353e+04, -2.14188081e+04,
       -7.21371109e+04, -1.11442426e+05, -8.68478612e+04, -7.65221324e+04,
       -9.31030277e+04, -5.06629450e+04, -4.73458775e+04, -7.26455025e+04,
       -1.07719043e+04, -6.79815582e+04, -1.01165429e+04,  2.47629686e+04,
        1.30442818e+05,  1.29337412e+05,  2.32621721e+05,  2.44191860e+05,
        4.35970621e+05,  5.89512044e+05,  5.02759802e+05,  3.07819907e+05,
       -6.48130760e+05,  7.82719410e+04, -3.93619294e+03, -9.85288719e+04,
        1.91616406e+05, -8.45665255e+05, -5.51706733e+04, -1.82279350e+04,
        3.27876688e+04, -4.12288675e+04])

In [14]:
pd.DataFrame({
    "Variables" : x.columns.values, 
    "Coefficients" : model.coef_}) 

Variables   Coefficients
0                      km_driven      -0.425282
1                      year_1995    8057.261691
2                      year_1996 -205140.128896
3                      year_1997 -382411.680965
4                      year_1998  -29024.529919
5                      year_1999  -46793.996189
6                      year_2000  -48078.535350
7                      year_2001  -21418.808072
8                      year_2002  -72137.110871
9                      year_2003 -111442.426128
10                     year_2004  -86847.861166
11                     year_2005  -76522.132385
12                     year_2006  -93103.027664
13                     year_2007  -50662.944989
14                     year_2008  -47345.877503
15                     year_2009  -72645.502511
16                     year_2010  -10771.904325
17                     year_2011  -67981.558162
18                     year_2012  -10116.542886
19                     year_2013   24762.968622
20                     year_2014  130442.817501
21                     year_2015  129337.411953
22                     year_2016  232621.721339
23                     year_2017  244191.859659
24                     year_2018  435970.620525
25                     year_2019  589512.044250
26                     year_2020  502759.802418
27                   fuel_Diesel  307819.907137
28                 fuel_Electric -648130.759571
29                      fuel_LPG   78271.940970
30                   fuel_Petrol   -3936.192941
31        seller_type_Individual  -98528.871879
32  seller_type_Trustmark Dealer  191616.406194
33           transmission_Manual -845665.255073
34    owner_Fourth & Above Owner  -55170.673317
35            owner_Second Owner  -18227.935008
36          owner_Test Drive Car   32787.668788
37             owner_Third Owner  -41228.867487

In [15]:
import statsmodels.api as sm 

In [16]:
x_i = sm.add_constant(x) 
x_i 

const  km_driven  year_1995  year_1996  year_1997  year_1998  year_1999  \
0       1.0      70000          0          0          0          0          0   
1       1.0      50000          0          0          0          0          0   
2       1.0     100000          0          0          0          0          0   
3       1.0      46000          0          0          0          0          0   
4       1.0     141000          0          0          0          0          0   
...     ...        ...        ...        ...        ...        ...        ...   
4335    1.0      80000          0          0          0          0          0   
4336    1.0      80000          0          0          0          0          0   
4337    1.0      83000          0          0          0          0          0   
4338    1.0      90000          0          0          0          0          0   
4339    1.0      40000          0          0          0          0          0   

      year_2000  year_2001  year_2002  ...  fuel_Electric  fuel_LPG  \
0             0          0          0  ...              0         0   
1             0          0          0  ...              0         0   
2             0          0          0  ...              0         0   
3             0          0          0  ...              0         0   
4             0          0          0  ...              0         0   
...         ...        ...        ...  ...            ...       ...   
4335          0          0          0  ...              0         0   
4336          0          0          0  ...              0         0   
4337          0          0          0  ...              0         0   
4338          0          0          0  ...              0         0   
4339          0          0          0  ...              0         0   

      fuel_Petrol  seller_type_Individual  seller_type_Trustmark Dealer  \
0               1                       1                             0   
1               1                       1                             0   
2               0                       1                             0   
3               1                       1                             0   
4               0                       1                             0   
...           ...                     ...                           ...   
4335            0                       1                             0   
4336            0                       1                             0   
4337            1                       1                             0   
4338            0                       1                             0   
4339            1                       1                             0   

      transmission_Manual  owner_Fourth & Above Owner  owner_Second Owner  \
0                       1                           0                   0   
1                       1                           0                   0   
2                       1                           0                   0   
3                       1                           0                   0   
4                       1                           0                   1   
...                   ...                         ...                 ...   
4335                    1                           0                   1   
4336                    1                           0                   1   
4337                    1                           0                   1   
4338                    1                           0                   0   
4339                    1                           0                   0   

      owner_Test Drive Car  owner_Third Owner  
0                        0                  0  
1                        0                  0  
2                        0                  0  
3                        0                  0  
4                        0                  0  
...                    ...                ...  
4335              

In [17]:
sm_model = sm.OLS(y, x_i).fit() 
print(sm_model.summary()) 

                            OLS Regression Results                            
Dep. Variable:          selling_price   R-squared:                       0.485
Model:                            OLS   Adj. R-squared:                  0.481
Method:                 Least Squares   F-statistic:                     106.7
Date:                Thu, 21 Jan 2021   Prob (F-statistic):               0.00
Time:                        23:20:20   Log-Likelihood:                -62301.
No. Observations:                4340   AIC:                         1.247e+05
Df Residuals:                    4301   BIC:                         1.249e+05
Df Model:                          38                                         
Covariance Type:            nonrobust                                         
                                   coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------
const           

**Interpretation:** The r-squared is exactly the same as the value I got in my model built in R. It is a small r2 of 0.485, indicating that the model can only predict 49% of the variance of selling price of cars. 